Deployment restructure

thehonker opened the pull request:
This PR:
- Moves wiki wwwroot to wwwroot/
- Pulls config out of wwwroot
- Rewrites Dockerfiles for
  - nginx
  - php-fpm
- Sets up autobuilds for the above
- Rewrites docker-compose files
- Restructures how config/envvars are handled

~~Once merged and autobuilds are confirmed working, we'll need to touch production `docker-compose.yml` again to switch from local builds to ghcr.io.~~

It does not:
- Encompass either of the other currently open PRs

thehonker committed:
mv wiki wwwroot

thehonker committed:
pull config and user uploads out of wwwroot

thehonker committed:
new docker, compose, and example env files

thehonker committed:
github workflows to autobuild both the nginx and php containers

thehonker committed:
update gitignore and readme

thehonker committed:
use https instead of git:// for npm module "load"

thehonker committed:
adjust dockerfiles to increase build caching

thehonker committed:
rename config.php to config.php.example, add ban.example.php and coorsponding volume mount

thehonker committed:
correct host in nginx.conf

thehonker committed:
install some image handling libs in the runtime container

thehonker committed:
rename ban.php.example

thehonker committed:
add --sql-mode="" to mariadb cmdline, more volume mounts

thehonker committed:
add --sql-mode="" to dev compose stack as well.

thehonker committed:
cleanup compose files, tag locally built images with "dev"

itsrachelfish reviewed (changes requested):
For some reason I'm unable to connect to the web server, it just says connection reset. I followed the steps in the readme and configured the environment variables. Ran docker compose and it built the containers. Docker PS shows the containers running. :thinking: 

![image](https://user-images.githubusercontent.com/1670549/223950787-9024be54-c3f7-4818-92a3-31be1c2c32a9.png)


![image](https://user-images.githubusercontent.com/1670549/223950256-0e6e0ec0-ea5f-4df9-a4a1-7b08f4754422.png)

thehonker committed:
properly tag the dev-php container

thehonker commented:
Resolved in [https://github.com/wetfish/wiki/pull/15/commits/8cc1db7e5935d911bc0fd9ed8186244de2c8978f](https://github.com/wetfish/wiki/pull/15/commits/8cc1db7e5935d911bc0fd9ed8186244de2c8978f), interesting behavior though.

thehonker committed:
change localhost port in dev stack

itsrachelfish commented:
@g-o-o-s FYI I rebased this PR to fix a merge conflict

itsrachelfish reviewed (changes requested):
It looks like debain stretch is no longer supported?

![Screenshot from 2023-06-03 01-47-34](https://github.com/wetfish/wiki/assets/1670549/73031570-798c-43a6-9c06-0b4d5622c721)
![Screenshot from 2023-06-03 01-47-08](https://github.com/wetfish/wiki/assets/1670549/77c50f52-4f55-4cc2-abb7-6912b70f2bf8)

thehonker committed:
rework for newer debian base image, other fixes
- rewrite dockerfiles to suit
- change how we pass in php-fpm configuration
- update workflows
- update/normalize example config
- improve readme a little bit
- add commented out bind mounts for wwwroot in dev envs

thehonker committed:
add watchtower for automated container updates

thehonker committed:
only expose dev stack on localhost

thehonker committed:
properly name workflows (^c^v lol)

thehonker commented:
> It looks like debain stretch is no longer supported?

:neat:

fd4508d rewrites the dockerfiles to base things on debian bookworm,
it also changes how we pass in php configuration and updates workflows to better match what we need.

thehonker committed:
correct dockerfile in workflow

thehonker committed:
correct dockerfile locations for both workflows (derp)

thehonker committed:
bump upload_max_filesize and post_max_size to 32M

itsrachelfish commented on Dockerfile.nginx:
```diff
@@ -0,0 +1,97 @@
+
+# acquire install-php-extensions script
+RUN set -exu \
+  && curl -fsSL https://github.com/mlocati/docker-php-extension-installer/releases/download/2.1.28/install-php-extensions \
```
Wow I never realized this was a whole thing. Why do php extensions not play nicely with docker? :thinking:

thehonker replied:
Mainline php docker images include helper scripts that manage php extensions along with required apt/apk packages, that script is meant to be included in those to add further functionality. We don't need it as we can get extensions from repos, and it actually refuses to work on a non-php-docker-mainline base image.

itsrachelfish commented on Dockerfile.nginx:
```diff
@@ -0,0 +1,97 @@
+
+# install php5.6, some extensions, and nodejs
+RUN set -exu \
+  && DEBIAN_FRONTEND=noninteractive apt-get -yq update \
+  && DEBIAN_FRONTEND=noninteractive apt-get -yq install \
+    php5.6 \
+    php5.6-fpm \
+    php5.6-mysqli \
+    php5.6-mysql \
+    php5.6-exif \
+    php5.6-gd \
```
If all these php extensions are being installed via apt-get, what is the `docker-php-extension-installer` doing exactly?

thehonker replied:
Nothing, good catch, it shouldn't be in there anymore.

itsrachelfish commented on Dockerfile.php:
```diff
@@ -0,0 +1,77 @@
+
+# Expose port 9000 and start php-fpm server
+EXPOSE 9000
+CMD ["/usr/sbin/php-fpm5.6", "--nodaemonize", "--force-stderr"]
```
Why `--force-stderr`? I'm assuming this makes php output errors to stderr instead of a log file. And `stderr` is automatically captured by the docker logs? Am I understanding this correctly?

thehonker replied:
Pretty much yeah:
```
--force-stderr
-O
    Force output to stderr in nodaemonize even if stderr is not a TTY.
```

itsrachelfish commented on mariadb.env.example:
```diff
@@ -0,0 +1,9 @@
+MARIADB_AUTO_UPGRADE=true
+MARIADB_ROOT_HOST=localhost
```
Shouldn't this be `%` instead of `localhost` to allow connections from database GUIs?

thehonker replied:
Sure should! Changed in 6fd33ab

itsrachelfish reviewed (changes requested):
I was able to get this to build :tada: 

Unfortunately it looks like the `README` needs to be updated to include a step about editing the `config/config.php` file. If you start the containers without creating this file you get a bunch of errors and a blank page:

![Screenshot from 2023-06-04 01-18-24](https://github.com/wetfish/wiki/assets/1670549/57028d05-af42-4180-b367-2d4aa4aa9e10)

After fixing the config file issue I tired restarting my containers and got a database error?

![Screenshot from 2023-06-04 01-26-05](https://github.com/wetfish/wiki/assets/1670549/407d55cf-8c23-454e-9926-b045268635cd)

itsrachelfish commented:
As per conversation on IRC:

- Enable live editing by default when using dev compose file
- Uncomment config.php from the gitignore
- Get rid of config.example.php
- Use getenv for all the constants defined in config.php

thehonker committed:
fixes:
- properly pass through db connection envvars to php-fpm workers
- expose mariadb on localhost in dev
- remove unused "install-php-extensions" script

thehonker committed:
fixes:
- properly passthrough DB_PASSWORD envvar to fpm workers
- properly order mounts in compose.dev.yml so config.php makes it in
- rm unnecessary install-php-extensions script from nginx build,
  rename fishy user to builder to better
  reflect actual usage

thehonker commented:
> * Enable live editing by default when using dev compose file

Done, [docker-compose.dev.yml#L43](https://github.com/wetfish/wiki/blob/deployment-restructure/docker-compose.dev.yml#L43)

> * Uncomment config.php from the gitignore
> * Get rid of config.example.php
> * Use getenv for all the constants defined in config.php

Not sure if we can do that for all the things in config.php, it has more than just db connection info in it.

As of [583aff0](https://github.com/wetfish/wiki/pull/15/commits/583aff086111777d26d5f2d6401e55695403e313) the site loads, however page creation/edits appear to be dropped on the floor without error.

thehonker committed:
move everything from config.php and ban.php into envvars

thehonker committed:
remove unnecessary recaptcha envvars

thehonker committed:
remove ban.php and config.php language from readme

thehonker committed:
normalize how we handle npm/node_modules, update readme to point out dev mode requirements

itsrachelfish reviewed (changes requested):
This is mostly working now, however I ran into an edge case with the "load" function.

Basically, the load function is used to fetch data from wetfish pages using php's `file_get_contents` function. It needs make an HTTP request to fetch the rendered content from the server, but `127.0.0.1:2405`  is only valid from outside the container. Inside the PHP container, it's unable to make the request and load the page content.

![image](https://github.com/wetfish/click/assets/1670549/152139d7-75e7-461c-a74a-c07c7c963d7d)

thehonker committed:
pass more envvars to php-fpm worker pools

itsrachelfish committed:
Refactored load[] function to handle internal docker hostnames

itsrachelfish committed:
Prevent redirect error when ?random displays the home page

itsrachelfish reviewed (dismissed):
I had to refactor how the `load[]` function handles URLs, but now these changes are officially working :tada: 

![Screenshot from 2023-06-10 02-30-47](https://github.com/wetfish/wiki/assets/1670549/2ae6eecd-63e1-4628-881d-337317c13a6a)

itsrachelfish reviewed (approved):
Made a small update to the example regex for allowed embeds.

In retrospect using a regex for matching a list of allowed domains is fragile and unsafe. It would be better to use a hardcoded array of valid domains.
